File: 199. Binary Tree Right Side View/main.js
var rightSideView = function (root) {
    if(!root) return [];

    let queue = [root];

    let res = [];

    while(queue.length){
        let levelsize = queue.length;
        for(let i=0; i<levelsize; i++){
            let current = queue.shift();
            if(i === levelsize-1){
                res.push(current.val)
            }
            if(current.left){
                queue.push(current.left);
            }
            if(current.right){
                queue.push(current.right);
            }
        }
    }
    return res;
};

/*
    1
    / \
    2   3
     \   \
      5   4
*/

const tree1 = {
    val: 1,
    left: {
        val: 2,
        left: null,
        right: {
            val: 5,
            left: null,
            right: null
        }
    },
    right: {
        val: 3,
        left: null,
        right: {
            val: 4,
            left: null,
            right: null
        }
    }
};

console.log(rightSideView(tree1));

File: 199. Binary Tree Right Side View/main.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []
        result = []
        queue = [root]
        while len(queue) > 0:
            len_q = len(queue)
            popcounter = 0
            for _ in range(len_q):
                n = queue.pop(0)
                popcounter+=1
                if popcounter == len_q:
                    result.append(n.val)
                if n.left is not None:
                    queue.append(n.left)
                if n.right is not None:
                    queue.append(n.right)

        return result
                


s = Solution()
# Test Case 1:
tree1 = TreeNode(1, TreeNode(2, None, TreeNode(5)), TreeNode(3, None, TreeNode(4)))
# Expected: [1,3,4]
print(s.rightSideView(tree1))

# Test Case 2:
tree2 = TreeNode(1, None, TreeNode(3))
# Expected: [1,3]
print(s.rightSideView(tree2))
